glean-core/rlb/src/private/datetime.rs

```diff
@@ -0,0 +1,134 @@
+pub struct DatetimeMetric(pub(crate) Arc<glean_core::metrics::DatetimeMetric>);
+
+impl DatetimeMetric {
+    /// The public constructor used by automatically generated metrics.
+    pub fn new(meta: glean_core::CommonMetricData, time_unit: TimeUnit) -> Self {
+        Self(Arc::new(glean_core::metrics::DatetimeMetric::new(
+            meta, time_unit,
+        )))
+    }
+}
+
+#[inherent(pub)]
+impl glean_core::traits::Datetime for DatetimeMetric {
+    fn set_with_details(
+        &self,
+        year: i32,
+        month: u32,
+        day: u32,
+        hour: u32,
+        minute: u32,
+        second: u32,
+        nano: u32,
+        offset_seconds: i32,
+    ) {
+        let metric = Arc::clone(&self.0);
+        dispatcher::launch(move || {
+            crate::with_glean(|glean| {
+                metric.set_with_details(
+                    glean,
+                    year,
+                    month,
+                    day,
+                    hour,
+                    minute,
+                    second,
+                    nano,
+                    offset_seconds,
+                )
+            })
+        });
+    }
+
+    fn set(&self, value: Option<Datetime>) {
+        let metric = Arc::clone(&self.0);
+        dispatcher::launch(move || crate::with_glean(|glean| metric.set(glean, value)));
+    }
+
+    fn test_get_value<'a, S: Into<Option<&'a str>>>(&self, ping_name: S) -> Option<Datetime> {
+        dispatcher::block_on_queue();
+
+        let queried_ping_name = ping_name
+            .into()
+            .unwrap_or_else(|| &self.0.meta().send_in_pings[0]);
+
+        crate::with_glean(|glean| self.0.test_get_value(glean, queried_ping_name))
+    }
+
+    fn test_get_num_recorded_errors<'a, S: Into<Option<&'a str>>>(
+        &self,
+        error: ErrorType,
+        ping_name: S,
+    ) -> i32 {
+        dispatcher::block_on_queue();
+
+        crate::with_glean_mut(|glean| {
+            glean_core::test_get_num_recorded_errors(&glean, self.0.meta(), error, ping_name.into())
+                .unwrap_or(0)
+        })
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use chrono::prelude::*;
+    use crate::common_test::{lock_test, new_glean};
+    use crate::CommonMetricData;
+
+    #[test]
+    fn datetime_convenient_api() {
+        let _lock = lock_test();
+        let _t = new_glean(None, true);
+
+        let metric: DatetimeMetric = DatetimeMetric::new(
+            CommonMetricData {
+                name: "datetime".into(),
+                category: "test".into(),
+                send_in_pings: vec!["test1".into()],
+                ..Default::default()
+            },
+            TimeUnit::Day,
+        );
+
+        // Record a date: it will get truncated to Day resolution.
+        metric.set_with_details(2018, 2, 25, 11, 5, 0, 0, 0);
+
+        // Check that the value has the correct resolution.
+        let date = metric.test_get_value(None).unwrap();
+        assert_eq!(
```

Comment:
This is currently weirdly failing because it says that the dates are different. However, printing both of them yields the same result: `2018-02-25T00:00:00+00:00`. Any clue what could be the culprit, @badboy ?

Reply:
Is it failing? On CI the next assert is failing only

Reply:
Ahhhh that's what happens when you stare at the same thing too much. And that's why it didn't make any sense in my brain :D